File: src/Random.cs
// MIT License 
// Copyright (C) 2019 Ara 3D. Inc
// https://ara3d.com

namespace Ara3D
{
    public static class StatelessRandom
    {
        public static uint RandomUInt(int index, int seed) 
            => (uint)Hash.Combine(seed, index);        

        public static int RandomInt(int index, int seed) 
            => Hash.Combine(seed, index);

        public static float RandomFloat(float min, float max, int index, int seed)
            => (float)RandomUInt(index, seed) / uint.MaxValue * (max - min)+ min;        

        public static float RandomNormalFloat(int index, int seed)
            => RandomFloat(0, 1, index, seed);        
    }
}

File: dart/lib/src/random.dart
// MIT License
// Copyright (C) 2019 VIMaec LLC.
// Copyright (C) 2019 Ara 3D. Inc
// https://ara3d.com

part of '../vim_math3d.dart';

class StatelessRandom {
  // static int randomUInt(int index, [int seed = 0]) {
  //   return Hash.combine(seed, index);
  // }
  static int randomInt(int index, [int seed = 0]) {
    return Hash.combine(seed, index);
  }

  static double randomFloatMinMax(double min, double max, int index,
      [int seed = 0]) {
    final hugeInt = double.maxFinite.toInt();
    return randomInt(index, seed) / hugeInt * (max - min) + min;
  }

  static double randomFloat(int index, [int seed = 0]) {
    return randomFloatMinMax(0.0, 1.0, index, seed);
  }

  static Vector2 randomVector2(int index, [int seed = 0]) {
    return Vector2(
      randomFloat(index * 2, seed),
      randomFloat(index * 2 + 1, seed),
    );
  }

  static Vector3 randomVector3(int index, [int seed = 0]) {
    return Vector3(
      randomFloat(index * 3, seed),
      randomFloat(index * 3 + 1, seed),
      randomFloat(index * 3 + 2, seed),
    );
  }

  static Vector4 randomVector4(int index, [int seed = 0]) {
    return Vector4(
        randomFloat(index * 4, seed),
        randomFloat(index * 4 + 1, seed),
        randomFloat(index * 4 + 2, seed),
        randomFloat(index * 4 + 3, seed));
  }
}
